xarray/core/ops.py

```diff
@@ -153,6 +153,22 @@ def fillna(data, other, join="left", dataset_join="left"):
                        dataset_fill_value=np.nan,
                        keep_attrs=True)
 
+def ffill(data, dim, limit=None):
+    axis_num = data.get_axis_num(dim)
+
+    if not has_bottleneck:
+        raise ImportError('ffill requires bottleneck to be installed')
+
+    data = data.copy()
+
+    if limit is None:
+        # bottleneck raises an error if you pass `None`
+        data.values = bn.push(data.values, axis=axis_num)
```

Comment:
Try `apply_ufunc` + `transpose()` afterwards instead? That would work on arbitrary xarray objects. It's awkward to need to restore the original dimension order, but it adds almost no overhead given NumPy's memory model. @jhamman is doing something like that in #1640.

I think you could even use `dask='parallelized'` here, to allow for doing multiple forward fills at the same time. 

Parallelizing across the pushed axis with dask would be harder (and certainly isn't necessary for a first pass), because you would need something hierarchical that can push data between chunks. But I think you could still make this work with multiple passes to get forward fill-values on the boundaries.